### PR TITLE
Customfilesystemmodel grandparent

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
This passes test scenario 10, but leaves 11 and 13 as failing.  I'm currently double-checking my solutions for 11 and 13, and will upload them to a separate PR in case you want to see everything at once.

**Comment by gperciva** [comment]
Err, wait a moment... why did I include `(int role)` in the definition of `::dataInteral()`?  Now I'm thinking that I should have removed that.

I won't change it yet, but if you have any thoughts on this I'd love to hear them.

File: src/customfilesystemmodel.cpp [hunk 1]
```diff
@@ -82,27 +110,48 @@ bool CustomFileSystemModel::setData(const QModelIndex &index,
...
-            if(parent.isValid())
+            // Recursively set PartiallyChecked on all ancestors, and make an
+            // ordered list of checked ancestors.
+            QModelIndex        parent         = index.parent();
+            QModelIndex        previousParent = index;
+            QList<QModelIndex> checkedAncestors;
+            while(parent.isValid())
             {
-                if(parent.data(Qt::CheckStateRole) == Qt::Checked)
+                if(hasCheckedAncestor(parent))
```
Comment (shinnok) [inline src/customfilesystemmodel.cpp]: Does this check current or parent? Am a bit confused.

**Comment by shinnok** [comment]
Regarding dataInteral(), depends on how you want to use it. If you only plan to use it in setIndexCheckState() I would remove role along with the asserts.

File: src/customfilesystemmodel.cpp (hunk 1 above)
Comment (gperciva) [inline src/customfilesystemmodel.cpp]: If `parent` is checked, or if any ancestor of `parent` is checked, it returns true.

I'm trying to think of a better name... `isCheckedOrAncestorChecked` is more clear, but it's too long.  Another option would be to write

    if((dataInternal(parent) == Qt::Checked) || (hasCheckedAncestor(parent))

and change `hasCheckedAncestor` to only check ancestors of the given index, but that also seems too verbose.

**Comment by gperciva** [comment]
After some consideration, I've gone with the slightly longer

    if((dataInternal(parent) == Qt::Checked) || (hasCheckedAncestor(parent))

line, since it's easier to understand what's happening.  If you like this adjusted PR, I'll rebase the two later commits to their previous ones.

File: src/customfilesystemmodel.cpp [hunk 2]
```diff
@@ -82,27 +104,49 @@ bool CustomFileSystemModel::setData(const QModelIndex &index,
...
+            {
+                if((dataInternal(parent) == Qt::Checked) ||
+                        hasCheckedAncestor(parent))
+                    checkedAncestors << parent;
+
+                // Set parent to be PartiallyChecked.
+                setIndexCheckState(parent, Qt::PartiallyChecked);
+
+                // Ascend to higher level of ancestor.
+                previousParent = parent;
```
Comment (shinnok) [inline src/customfilesystemmodel.cpp]: The changes are good, however do you really need previousParent in this first block? I think you only need it further down for the sibling iteration.

**Comment by gperciva** [comment]
Right, fixed.  I also re-ran `clang-format`.

**Comment by shinnok** [comment]
Looks good, do you want me to merge this now or test it in parallel with the changes for the other failing cases?

**Comment by gperciva** [comment]
Rebased history.  Please merge this now; I'll put the other fixes separately.

**Pull request merged**